File: code-samples/messaging/quick-start.js
const RC = require('@ringcentral/sdk').SDK
const path = require('path')
// Remember to modify the path to where you saved your .env file!
require('dotenv').config({ path: path.resolve(__dirname, '../.env') })

// For the purpose of testing the code, we put the SMS recipient number in the environment variable.
// Feel free to set the SMS recipient directly.
const RECIPIENT    = process.env.SMS_RECIPIENT

// Instantiate the SDK and get the platform instance
var rcsdk = new RC({
    'server':       process.env.RC_SERVER_URL,
    'clientId':     process.env.RC_CLIENT_ID,
    'clientSecret': process.env.RC_CLIENT_SECRET
});
var platform = rcsdk.platform();

/* Authenticate a user using a personal JWT token */
platform.login({ 'jwt':  process.env.RC_JWT })

platform.on(platform.events.loginSuccess, function(e){
    read_extension_phone_number_detect_sms_feature()
});

platform.on(platform.events.loginError, function(e){
    console.log("Unable to authenticate to platform. Check credentials.", e.message)
    process.exit(1)
});

/*
  Read phone number(s) that belongs to the authenticated user and detect if a phone number
  has the SMS capability
*/
async function read_extension_phone_number_detect_sms_feature(){
    try {
        let endpoint = "/restapi/v1.0/account/~/extension/~/phone-number"
        var resp = await platform.get(endpoint)
        var jsonObj = await resp.json()
        for (var record of jsonObj.records){
            for (var feature of record.features){
                if (feature == "SmsSender"){
                    // If a user has multiple phone numbers, check and
		    // decide which number to be used for sending
		    // the SMS message. For simplicity, we pick the
		    // first one we find. 
                    return send_sms(record.phoneNumber)
                }
            }
        }
        if (jsonObj.records.length == 0)
          console.log("This user does not own a phone number!")
        else
          console.log("None of this user's phone number(s) has the SMS capability!")
    } catch(e) {
        console.log(e.message)
        process.exit(1)
    }
}

/*
 Send a text message from a user own phone number to a recipient number
*/
async function send_sms(fromNumber){
    try {
        let bodyParams = {
            from: { phoneNumber: fromNumber },
            to: [ { phoneNumber: RECIPIENT} ],
            // To send group messaging, add more (max 10 recipients) 'phoneNumber' object. E.g.
            /*
            to: [
               { phoneNumber: RECIPIENT },
               { phoneNumber: 'Recipient-Phone-Number' }
             ],
            */
            text: 'Hello World!'
        }
        let endpoint = "/restapi/v1.0/account/~/extension/~/sms"
        var resp = await platform.post(endpoint, bodyParams)
        var jsonObj = await resp.json()
        console.log("SMS sent. Message id: " + jsonObj.id)
        check_message_status(jsonObj.id);
    } catch(e) {
        console.log(e.message)
        process.exit(1)
    }
}

/*
 Check the sending message status until it's out of the queued status
*/
async function check_message_status(messageId){
    try {
        let endpoint = `/restapi/v1.0/account/~/extension/~/message-store/${messageId}`
        let resp = await platform.get(endpoint);
        let jsonObj = await resp.json()
        console.log("Message status: ", jsonObj.messageStatus)
        if (jsonObj.messageStatus == "Queued"){
          await sleep (5000);
          check_message_status(jsonObj.id);
        }
    } catch (e) {
      console.log(e.message)
      process.exit(1)
    }
}

const sleep = async (ms) => {
  await new Promise(r => setTimeout(r, ms));
}

/**********************************************************
 Code snippet section for boostrap testing purpose
**********************************************************/
// For option 1
const RECIPIENT2    = process.env.SMS_RECIPIENT2
exports.platform = platform;
exports.RECIPIENT = RECIPIENT;
exports.RECIPIENT2 = RECIPIENT2;

//boostrap_test_function()

async function boostrap_test_function(){
/*
  await sleep(2000)
  console.log("Read SMS feature")
  require ('./code-snippets/number-features.js')

  await sleep(2000)
  console.log("Test sending MMS")
  require ('./code-snippets/send-mms.js')

  await sleep(2000)
  console.log("Test sending Fax")
  require ('./code-snippets/send-fax.js')

  await sleep(2000)
  console.log("Read message store")
  require ('./code-snippets/message-store.js')

  await sleep(2000)
  console.log("Export message store")
  require ('./code-snippets/message-store-export.js')

  await sleep(2000)
  console.log("Test sending HV SMS")
  require ('./code-snippets/send-a2p-sms.js')
*/
  await sleep(2000)
  console.log("Test access HV message store")
  require ('./code-snippets/a2p-message-store.js')
}

File: code-samples/messaging/quick-start.py
import os, sys, time
from dotenv import load_dotenv
from ringcentral import SDK

load_dotenv()

# Read phone number(s) that belongs to the authenticated user and detect if a phone number
# has the SMS capability
def read_extension_phone_number_detect_sms_feature():
  try:
    resp = platform.get("/restapi/v1.0/account/~/extension/~/phone-number")
    jsonObj = resp.json()
    for record in jsonObj.records:
      for feature in record.features:
        if feature == "SmsSender":
          # If user has multiple phone numbers, check and decide which number
          # to be used for sending SMS message. For simplicity, we pick the first one we find.
          return send_sms(record.phoneNumber)
    if count(jsonObj.records) == 0:
      print("This user does not own a phone number!");
    else:
      print("None of this user's phone number(s) has the SMS capability!");
  except Exception as e:
    print (e.message)

# Send a text message from a user own phone number to a recipient number
def send_sms(fromNumber):
  try:
    bodyParams = {
         'from' : { 'phoneNumber': fromNumber },
         'to'   : [ {'phoneNumber': RECIPIENT} ],
         # To send group messaging, add more (max 10 recipients) 'phoneNumber' object. E.g.
         #
         #'to': [
         #       { 'phoneNumber': "Recipient1-Phone-Number" },
         #       { 'phoneNumber': "Recipient2-Phone-Number" }
         #     ],
         'text' : "Hello World!"
    }
    endpoint = "/restapi/v1.0/account/~/extension/~/sms"
    resp = platform.post(endpoint, bodyParams)
    jsonObj = resp.json()
    print ("SMS sent. Message id: " + str(jsonObj.id))
    check_message_status(jsonObj.id)
  except Exception as e:
    print (e)

# Check the sending message status until it's out of the queued status
def check_message_status(messageId):
  try:
    endpoint = "/restapi/v1.0/account/~/extension/~/message-store/" + str(messageId)
    resp = platform.get(endpoint)
    jsonObj = resp.json()
    print ("Message status: " + jsonObj.messageStatus)
    if (jsonObj.messageStatus == "Queued"):
      time.sleep(2)
      check_message_status(jsonObj.id)
  except Exception as e:
    print (e)

# Instantiate the SDK and get the platform instance
rcsdk = SDK( os.environ.get('RC_APP_CLIENT_ID'),
             os.environ.get('RC_APP_CLIENT_SECRET'),
             "https://platform.ringcentral.com" )
platform = rcsdk.platform()

# For the purpose of testing the code, we put the SMS recipient number in the environment variable.
# Feel free to set the SMS recipient directly.
RECIPIENT    = os.environ.get('SMS_RECIPIENT')

# Authenticate a user using a personal JWT token
def login():
    try:
      platform.login( jwt=os.environ.get('RC_USER_JWT') )
      read_extension_phone_number_detect_sms_feature()
    except Exception as e:
      sys.exit("Unable to authenticate to platform. Check credentials." + str(e))

login()

##############################
# End of quick start section
##############################

########
# Code snippet section for boostrap testing purpose
########
RECIPIENT2    = os.environ.get('SMS_RECIPIENT2')

def boostrap_test_function():
  import importlib

  # time.sleep(2)
  # print ("Test reading number feature")
  # ms = importlib.import_module("code-snippets.number-features")
  # ms.platform = platform
  # ms.json = json
  # ms.detect_sms_feature()

  # print ("Test sending MMS")
  # time.sleep(2)
  # mms = importlib.import_module("code-snippets.send-mms")
  # mms.platform = platform
  # mms.rcsdk = rcsdk
  # mms.time = time
  # mms.RECIPIENT = RECIPIENT
  # mms.read_extension_phone_number_detect_mms_feature()
  #
  # time.sleep(2)
  # print ("Test sending Fax")
  # fax = importlib.import_module("code-snippets.send-fax")
  # fax.platform = platform
  # fax.rcsdk = rcsdk
  # fax.time = time
  # fax.RECIPIENT = RECIPIENT
  # fax.send_fax()
  #
  # time.sleep(2)
  # print ("Test reading message store")
  # ms = importlib.import_module("code-snippets.message-store")
  # ms.platform = platform
  # import json
  # ms.json = json
  # ms.read_extension_message_store()

  # time.sleep(2)
  # print ("Test export message store")
  # ms = importlib.import_module("code-snippets.message-store-export")
  # ms.platform = platform
  # ms.create_message_store_report()

  # time.sleep(2)
  # print ("Test sending a2p sms batch")
  # ms = importlib.import_module("code-snippets.send-a2p-sms")
  # ms.platform = platform
  # ms.RECIPIENT = RECIPIENT
  # ms.RECIPIENT2 = RECIPIENT2
  # ms.read_extension_phone_number_detect_a2psms_feature()


  time.sleep(2)
  print ("Test receive reply sms")
  ms = importlib.import_module("code-snippets.receive-reply-sms")
  ms.platform = platform
  ms.rcsdk = rcsdk
  ms.login()

########
# boostrap_test_function()
